contents/blog/how-we-turned-clickhouse-into-our-eventmansion.md

```diff
@@ -0,0 +1,84 @@
+
+
+## Choosing ClickHouse
+
+Faced with this need, we started investigating replacements for Postgres. We looked at a wide range of OLAP solutions, including Pinot, Presto, Druid, TimescaleDB, CitrusDB and ClickHouse. Some of our team had used these tools before at other companies, such as Uber where Pinot and Presto are both used extensively. 
+
+We assessed each tool we looked at on three main factors:
+
+- *Speed:* Our users want results in real-time, so our new database needed to scale really well and give fast results. Ideally it wouldn’t be too expensive either.
+- *Complexity:* PostHog users can self-host and install our product themselves, so we didn’t want it to be too complicated for users to manage or deploy. We didn’t want users to have to install an entire Hadoop stack, for example.
+- *Query interface:* We like standardised tools. We eliminated tools such as Druid because, while it does have a SQL wrap around it, it’s not _exactly_ SQL. That can get messy. 
+
+ClickHouse was a good fit for all of these factors, so we started doing a more thorough investigation. We read up on benchmarks and researched the experience of companies such as [Cloudflare that uses ClickHouse to process 6m requests per second](https://blog.cloudflare.com/http-analytics-for-6m-requests-per-second-using-clickhouse/). Next, we set up a test cluster to run our own benchmarks. 
+
+ClickHouse repeatedly performed an order of magnitude better than other tools we considered and we discovered other perks, such as the fact that it is column-orientated and written in C++. 
+
+- *Compression:* ClickHouse has _excellent_ compression and the size-on-disk was incredible. ClickHouse even beat out serialization formats such as ORC and Parquet. 
+- *Process from disk:* Some OLAP solutions, like Presto, require data to live in memory. That’s fast, but you need to have a _lot_ of memory for big datasets. ClickHouse processes from disk, which is better for smaller instances too. 
+- *Real-time data updates:* ClickHouse basically processes data as it arrives, so there’s no need to pre-aggregate data. It’s faster for us, and our users.
+
+Eventually, we decided we knew enough to proceed and so we spun our test cluster out into an actual production cluster. It’s just part of how [we like to bias for speed](https://posthog.com/careers). 
+
+## Turning ClickHouse into our EventMansion
+
+It wasn’t all smooth sailing. Mutations in particular tripped us up, as we initially used them a lot instead of taking the time to model the data perfectly for ClickHouse. The documentation suggests not running more than one or two mutations an hour. We were running several hundred per minute. 
```

Comment:
We should link to the mutations docs. This? https://clickhouse.com/docs/en/sql-reference/statements/alter/#mutations